syncer/pipe.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"sync"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+)
+
+// Pipe is the littlest process unit in syncer
+type Pipe interface {
+	// Name is the pipe's name
+	Name() string
+
+	// Init initializes the pipe
+	Init(*config.SubTaskConfig, func() error) error
```

Comment:
any explanation about this `func() error`?

Reply:
https://github.com/pingcap/dm/pull/160/commits/a5450bcd2bbc1336d43d6621f757b0f1608b4c53

Reply:
so, why call this extra init _func_ in the caller?

Reply:
I think this is also a part of initial work, should do in Init func

Reply:
IMO, there is an `Init` in `Syncer` which calls all `init`s of Pipes. but maybe your choice is better and we can keep considering this later.